feat: Add draft for post on SCv2 Hodometer

agrski opened the pull request:
# Why
## Motivation
Hodometer is one of the components in Seldon Core v2.  Any series of blog posts on Core v2 should therefore discuss it: what it is, how it works, why it was designed the way it was and any history or context informing those decisions, and so on.

I've chosen to start by drafting a post on Hodometer because I probably know it and the design decisions the best, and because it's a little independent of most of the rest of the system, thus there shouldn't be too many cross-references to iron out.

## Issues
N/A

# What
## Changes
* Add draft blog post about the SCv2 component Hodometer.

## Misc
In case it's easier to read and review, here is the link to the rendered Markdown: https://github.com/agrski/agrski.github.io/blob/add-draft-for-hodometer-post/docs/_drafts/scv2-hodometer.md

Certain formatting might look odd, like `---`, because that'll be rendered by Jekyll but not Markdown.

agrski committed:
Add notes for draft post on SCv2 Hodometer

agrski committed:
Add intro on purpose & origin of name of Hodometer

agrski committed:
Add sentence defining structure/sections of blog post as overview

agrski committed:
Add intro to Spartakus project

agrski committed:
Add paragraph on design goals of Spartakus

agrski committed:
Add paragraph on data collection endpoints in Spartakus & usage thereof by Seldon

agrski committed:
Add paragraph on problems with Spartakus

agrski committed:
Add link to Spartakus repo

agrski committed:
Add paragraph on key differences of Core v2 from Core v1

agrski committed:
Add list of desirable design decisions from Spartakus

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,68 @@
+## Designing Hodometer
+
+Core v2 is a very different kettle of fish from Core v1, and the sorts of metrics we'd want are no exception.
+Core v2 has multiple Kubernetes custom resources --- models, servers, pipelines, and experiments --- compared to just single one in Core v1 --- deployments.
```
:thought_balloon: Consider using the exact terms for resources, e.g. `SeldonDeployment`

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/70a3df4d13eb3bae8ee5fe5096002274f15be2ac and https://github.com/agrski/agrski.github.io/pull/40/commits/be55202a0abc2726c49e7abf95e95f599ee1f0dc

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,68 @@
+* Low periodicity reporting of metrics.
+* Push-based reporting of metrics.
+* Flexibility to configure additional metrics receivers.
+
```
:thought_balloon: Consider adding quoted text to emphasise specific design considerations and/or their impact

agrski replied:
For context, other blogs and articles use quotes as a form of call-out to emphasise specific sentences from the text

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/be4b17b10133bf35dac2b82ea592aaa83176e137

agrski commented on docs/_drafts/scv2-hodometer.md:
:thought_balloon: Consider adding Mermaid diagrams explaining how various things communicate, e.g. to contrast Spartakus & Hodometer, and to explain Hodometer's internal architecture

agrski replied:
Diagrams added

agrski committed:
Add explanations of desirability for 3 characteristics of Spartakus

agrski committed:
Fix typo

agrski committed:
Rephrase sentence on non-collection of PII for clarity

agrski committed:
Add explanation of desirability for push-based metrics

agrski committed:
Add detail re. export of metrics from OSS tool to third party

agrski committed:
Add explanation of desirability of low-periodicity metrics

agrski committed:
Reorder list of inspirations from Spartakus to match following paragraphs

agrski committed:
Fix error stating Spartakus supported additional receivers

As far as scanning the documentation and goes shows, Spartakus only supports one endpoint being configured at a time, but does allow this to be configured.

agrski committed:
Add paragraph on config of extra metrics receivers for Hodometer

agrski committed:
Add paragraph on metrics levels

agrski committed:
Split section on designing Hodometer into inspiration from Spartakus & new ideas

ukclivecox reviewed (commented):
Seems good but maybe if this was in a blog I assume you would also be addressing a wider audience with no interest in SCV2 so might be "tips and tricks for providing usage monitoring for your open source project" or something?

agrski committed:
Add paragraph on how to disable Hodometer

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,127 @@
+Resource-level metrics count the number of each Core v2-defined _custom resource_, such as the number of models and pipelines.
+It's important to note here that sensitive information like model names is _not_ recorded.
+Feature-level metrics are about how many servers have multi-model serving and over-committing enabled and how much memory servers have available to them.
+All of this can be seen in the `Collect` method [here](https://github.com/SeldonIO/seldon-core/blob/d3502062bbbb18a08032201917ceea07e124be41/hodometer/pkg/hodometer/collect.go#L227).
```
:wrench: Should also link to `metrics.go` for the metrics definitions.  The `Collect` method best shows that levels are supersets/subsets of one another.

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/7c515e074dd451549955c4ca18232b5c118ec76f

agrski committed:
Add paragraph on design goal of simplicity re. metrics

agrski committed:
Remove points from notes at the end which have been covered

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,135 @@
+Spartakus does support some custom data in the form of [extensions](https://github.com/kubernetes-retired/spartakus/tree/master/docs#extensions), but this is static information.
+In any case, users may not even want to share all the data on nodes in a cluster that Spartakus collects by default.
+To put it concisely, Spartakus collected lots of data that _wasn't_ of interest but didn't collect lots of data that _was_ of interest!
+
```
:thought_balloon: Consider adding a system-level diagram for Spartakus to use as a point of comparison for ones about Hodometer later

agrski replied:
Diagrams added for both Spartakus & Hodometer at the container and also component levels

agrski replied:
May still want to add a link to it here, but perhaps it's fine to have that later

agrski commented:
> Seems good but maybe if this was in a blog I assume you would also be addressing a wider audience with no interest in SCV2 so might be "tips and tricks for providing usage monitoring for your open source project" or something?

Interesting suggestion!  The intent for this blog post (I realise I haven't added a PR description yet) is to be about SCv2 specifically, as part of a series of deep-dives into the design and characteristics of its components and the system as a whole.

With that said, it seems like a good suggestion to have a summary at the end about things to consider if designing a similar metrics system for some other project, e.g. push vs. pull data collection, if support for different data stores is important, collection levels, and so on.

It'd also be worth having a bit more of an introduction to those topics towards the start--the good old "tell 'em what you're gonna tell 'em"!

agrski committed:
Add heading for section on Hodometer architecture

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,134 @@
+When you consider these things, it should be clear that the metrics for Core v2 will want to be correspondingly more granular and domain-focused.
+
+While Spartakus didn't provide ideal metrics even for Core v1, it did make a number of sensible design decisions to use as a basis for implementing usage metrics in Core v2.
+Specifically, these were:
```
:thought_balloon: These might be better handled as level three headings for the paragraphs below.  That would emphasise them better whilst simultaneously splitting the following content up into more visually manageable chunks.

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/ebd6829836084fcbd575c2feb3b50928b18546e8

agrski committed:
Capitalise names of SCv2 custom resources & italicise for emphasis as introducing

Many typesetting guidelines, such as those used by OReilly, emphasise new technical terminology.
Using italics for this purpose seems reasonable, although emboldening these terms would work too.

agrski committed:
Capitalise & italicise SCv1 custom resource name for emphasis

agrski committed:
Replace list of characteristics with section headings for explanatory paragraphs

agrski committed:
Split up explanatory paragraphs on Spartakus influences

Some of the existing paragraphs were rather long and contained multiple different ideas or avenues of consideration.
Smaller paragraphs are generally easier to read and maintain visual position in.

agrski committed:
Split up paragraph on problems with Spartakus

agrski committed:
Add headings & split paragraphs on new features over Spartakus

agrski committed:
Add intro to C4 container view of Hodometer architecture

agrski committed:
Add description of high-level behaviour of Hodometer

This will complement the visual insights of the Mermaid diagram (once created and referenced!) with a written explanation.

agrski committed:
Fix mistake calling container diagram component diagram

agrski committed:
Add nodes/actors in Spartakus container diagram

agrski committed:
Add blank lines in node texts to separate titles & descriptions

agrski committed:
Add links between nodes

The collector-volunteer arrow shows the direction of the dependency rather than the communication.
This is because I want the the consuming system on the opposite side to the dependencies of the Spartakus volunteer.
I might revisit this, as it doesn't reflect the active/passive sides of the comms and direction of information transfer.

agrski committed:
Add styling to nodes in Spartakus container diagram

agrski committed:
Add PNG of Spartakus container diagram

agrski committed:
Relabel & reflect volunteer/receiver edge for comms direction

agrski committed:
Reorder edges so deps on left and consumers on right of diagram

agrski committed:
Update Spartakus container diagram

agrski committed:
Add subgraph for services in k8s cluster in Spartakus arch

agrski committed:
Regenerate PNG of Spartakus container diagram

agrski committed:
Use concise node label syntax in Spartakus Mermaid diagram

agrski committed:
Add nodes/actors in Hodometer container diagram

agrski committed:
Add links between nodes to Hodometer container diagram

agrski committed:
Add styling to nodes in Hodometer container diagram

agrski committed:
Adjust ranks of receiver edges in Hodometer container diagram

agrski committed:
Add network protocols to edges in Hodometer container diagram

agrski committed:
Adjust rank of collector edge in Spartakus container diagram

agrski committed:
Add network protocols to edges in Spartakus container diagram

agrski committed:
Regenerate PNG of Spartakus container diagram

agrski committed:
Add PNG of Hodometer container diagram

agrski committed:
Remove redundant naming from node descriptions in Spartakus container diagram

agrski committed:
Regenerate PNG of Spartakus container diagram

agrski committed:
Update receiver node names in Hodometer container diagram for consistency with Spartakus equivalent

agrski committed:
Regenerate PNG of Hodometer container diagram

agrski committed:
Add style for k8s cluster subgraph in Spartakus container diagram

This style is a light grey background instead of of cream/yellow and features a dashed instead of solid border.

agrski committed:
Add style for k8s cluster subgraph in Hodometer container diagram

This matches the style for the Spartakus k8s cluster subgraph.

agrski committed:
Regenerate PNGs of Hodometer & Spartakus container diagrams

agrski committed:
Add Mermaid markup for Hodometer component diagram

agrski committed:
Add PNG of Hodometer component diagram

agrski committed:
Add Mermaid markup for Spartakus component diagram

agrski committed:
Add PNG of Spartakus component diagram

agrski commented:
@ukclivecox Would you mind having a squizz at the diagrams (the PNGs) and confirming those look reasonable?  As a brief explanation, the _container_ diagrams are about deployable units, and the _component_ diagrams are roughly analogous to services/interfaces/modules; they don't reflect how things are wired up and instantiations of those interfaces, as that's what the _code_ level in the C4 model is for.

I need to write it up still, but actually looking into this and creating these diagrams has brought up some interesting similarities and differences between Spartakus and Hodometers that I hadn't noticed before.

agrski committed:
Add link to Hodometer container diagram

agrski committed:
Add sentence on comms protocols used by Hodometer

agrski committed:
Add paragraph on metrics receivers for Hodometer

agrski committed:
Add paragraph on high-level arch of Spartakus re. diagram

agrski committed:
Add paragraphs on Spartakus component-level arch

agrski committed:
Add paragraph on Database interface in Spartakus component arch

agrski committed:
Add paragraphs on main interfaces/structs in Hodometer component arch

agrski committed:
Update arch info to reference both C4 levels depicted

agrski committed:
Add section headings for C4 arch views across Spartakus & Hodometer

agrski committed:
Add link to Hodometer component diagram

agrski committed:
Move inclusion of Hodometer component diagram to after intro to it

agrski committed:
Remove resolved comments on what to cover

agrski committed:
Conjoin sentences re. Collector impl using semi-colon

agrski committed:
Add justification of gRPC use re. Core v2 scheduler

agrski committed:
Remove done TODO comment

agrski committed:
Add conclusions section

agrski committed:
Remove introductory sentence to conclusions as superfluous

agrski committed:
Remove more resolved comments on what to cover

agrski commented on docs/_drafts/scv2-hodometer.md:
```diff
@@ -0,0 +1,199 @@
+
+Hodometer is an optional component of Seldon Core v2 responsible for collecting anonymous usage metrics.
+The name comes from the [ancient device](https://en.wikipedia.org/wiki/Odometer) for measuring distance, familiar in cars as an _odometer_, because it's all about keeping track of your mileage!
+
```
:wrench: Add section on who might be interested in this article, e.g. those looking to understand Core v2 in greater depth, those looking to implement usage metrics in a responsible manner for their own projects, and those simply looking to understand the design decisions in an (until recently) open-source project.

agrski replied:
@ukclivecox Once I've resolved this comment, that should also address your own previous comment.  Aside from that, I think this nearing completion now as a draft.  (A draft in the Jekyll sense, i.e. that it isn't "published" on the blog, not that it isn't in a good state.)

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/1ece90a8c3c9ec37389da49be94c16d5d946adc0

agrski commented on docs/_drafts/scv2-hodometer-hodometer-component.png:
:thought_balloon: Investigate adding copyright notices to the images, as these are liable to being reused by referencing articles and search engines, which may or may not include proper attribution

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/17e59d6d62ad66c4b89d3ecc18179f03f6e0a8fc

For reference, although I'll likely end up adding this into a Makefile at some point for convenience, I used the following bash script to add copyright notices:
```bash
find docs/_drafts/ -name '*.png' | while read -r F; do convert $F -pointsize 12 -fill grey50 label:"Copyright (C) agrski 2024" -gravity center -append $F; done
```

agrski committed:
Add copyright notices to Mermaid-generated diagrams using ImageMagick

agrski committed:
Update container arch diagram titles to specify C4 level

agrski committed:
Regenerate PNGs of Hodometer & Spartakus C4 diagrams

agrski committed:
Add more tags to post

agrski commented on docs/_drafts/scv2-hodometer-spartakus-container.mmd:
```diff
@@ -0,0 +1,23 @@
+  class C consumingSystem
+  class K supportingSystem
+
+  style k8s-cluster fill:#eee,stroke:#ccc,stroke-width:2px,stroke-dasharray: 5 5
```
:question: Can subgraph styles be turned into a `classDef`?  If so, that would slightly simplify the definitions by providing better grouping and supporting multiple subgraphs (if required in the future) without duplication

agrski replied:
The answer is a resounding (and fortunately unsurprising) yes.  Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/586cee78924a3b247ee4fe4734c9a81882f4849d

agrski committed:
Use class defs for subgraph styles in Mermaid diagrams

agrski committed:
Use condensed assignment of nodes to classes in Hodometer container diagram

agrski committed:
Regenerate PNGs of Hodometer & Spartakus C4 diagrams

agrski committed:
Fix typos

agrski committed:
Add link to metrics structs & update wording re. Collect() method

agrski committed:
Explain C4 model briefly + state contrasting Hodometer & Spartakus

agrski committed:
Fix unclosed backticks

agrski committed:
Add link to k8s client wrapper struct in Spartakus

agrski committed:
Add code snippets for the key logic of Hodometer's main components

agrski committed:
Minor rewordings for legibility, sense, and flow

agrski committed:
Emphasise Hodometer was consciously designed differently from Spartakus but ended up similar architecturally

agrski committed:
Fix indentation issue in code snippet

agrski committed:
Add set of rhetorical questions to conclusions

agrski committed:
Add ending paragraph on respect for your (online) community

agrski committed:
Add quotes from the text to highlight ideas

agrski committed:
Add section in introduction re. potential audience profiles

This suggestion from Clive should help readers quickly gauge if this article could be of interest to them.
While the focus is on Hodometer and Spartakus, there are interesting take-aways for a wider audience too.

agrski commented:
One final point to consider is ensuring consistency around formatting, i.e. the semantics for preformatted/monospace text, italics, and bold-face:
* Monospace should be for elements taken directly from code, e.g. file, struct, and method names.
* Bold should be used to highlight key terms that are not part of a code-base, e.g. the names of C4 diagram levels.
* Italics should be used for general emphasis in all other contexts, e.g. contrasting trade-offs between systems.

---

Edit: fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/b5587744a560143310c1ba05cf81a0c761d9f15c, https://github.com/agrski/agrski.github.io/pull/40/commits/16ea0e98dcda0ae6693bfc44df6e7e90bafea6d0, and https://github.com/agrski/agrski.github.io/pull/40/commits/4ba5c118922bf76fd3d0ae4fcb732405ff8afd59.

agrski committed:
Add sentence in introduction explaining article is part of series

This provides a convenient place, right at the start, to link to the GitHub project.

agrski committed:
Use backticks for k8s CRD names

agrski committed:
Replace italics with bold-face for technical terms & key concepts

agrski committed:
Use bold-face for name of C4 arch levels

agrski committed:
Minor rewording

agrski committed:
Remove emphasis for slice as Go version of a list

agrski committed:
Add blank lines to separate long paragraphs re. Hodometer's components

agrski committed:
Move subordinate clause re. MixPanel format to main sentence re. this

agrski commented on docs/_drafts/scv2-hodometer-hodometer-container.mmd:
```diff
@@ -0,0 +1,28 @@
+  classDef consumingSystem  fill:#08427b,stroke:#052e56,color:#ffffff
+  classDef systemBoundary   fill:#eee,stroke:#ccc,stroke-width:2px,stroke-dasharray: 5 5
+
+  class H,S,K focusSystem
```
:wrench: K and S are supporting systems, not focus systems

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/86476803ccb4c3acbc9cd65ff33e166b79e1007d

agrski committed:
Fix wrong class for supporting systems in Hodometer container diagram

agrski committed:
Regenerate PNG of Hodometer container diagram

agrski commented on docs/_drafts/scv2-hodometer.md:
````diff
@@ -0,0 +1,313 @@
+Instead, in Hodometer this is provided by the wiring logic in the `main` function, which can be found [here](https://github.com/SeldonIO/seldon-core/blob/d3502062bbbb18a08032201917ceea07e124be41/hodometer/cmd/hodometer/main.go#L70):
+
+```go
+	punctuator.Run(
````
:wrench: Although correctly indented for where it was copied from the code, the indentation in these code blocks isn't important for readers of the article.  In fact, given the rendering of tabs can be significant (8 spaces in GitHub), it probably hurts legibility to leave the tabs in place.

As such, the tabs in code blocks should be removed.

agrski replied:
Fixed in https://github.com/agrski/agrski.github.io/pull/40/commits/748e359c0099285a39080342694949c6478a6da0

agrski committed:
Remove leading indentation for code excerpts
